### PR TITLE
DEP: deprecation warnings for ``special.lpn`` and ``[c]lpmn``

### PR DESCRIPTION
**Pull request by jorenham**
ref: https://github.com/scipy/scipy/pull/22099#issuecomment-2555825723
follow-up of: https://github.com/scipy/scipy/pull/20539

The ``special.lpn`` and ``special.[c]lpmn`` functions should be deprecated since 1.15.0, as was noted in their docstrings, but they did not emit a `DeprecationWarning` when called at runtime.

Now they do.

### CODE REVIEW

**Comment by jorenham** [comment]
ping @h-vetinari

**Review by h-vetinari — APPROVED** [review]
Thanks, LGTM! 🙏 

CC @steppi @j-bowhay

**Comment by steppi** [comment]
Thanks @jorenham. This one's my fault. If I remember right, I had told @izaid I would add them in a follow-up to his Legendre function PR.

A couple things: 

1. We also need a deprecation warning for [`sph_harm`](https://scipy.github.io/devdocs/reference/generated/scipy.special.sph_harm.html#scipy.special.sph_harm), and suppression in the relevant tests.
2. Some tests in `test_mpmath.py` need to have the warning suppression added.

```
 =========================== short test summary info ============================
FAILED scipy/special/tests/test_mpmath.py::TestSystematic::test_legenp - DeprecationWarning: `scipy.special.lpmn` is deprecated as of SciPy 1.15.0 and will be removed in SciPy 1.17.0. Please use `scipy.special.assoc_legendre_p_all` instead.
FAILED scipy/special/tests/test_mpmath.py::TestSystematic::test_legenp_complex_2 - DeprecationWarning: `scipy.special.clpmn` is deprecated as of SciPy 1.15.0 and will be removed in SciPy 1.17.0. Please use `scipy.special.assoc_legendre_p_all` instead.
FAILED scipy/special/tests/test_mpmath.py::TestSystematic::test_legenp_complex_3 - DeprecationWarning: `scipy.special.clpmn` is deprecated as of SciPy 1.15.0 and will be removed in SciPy 1.17.0. Please use `scipy.special.assoc_legendre_p_all` instead.
= 3 failed, 74062 passed, 4634 skipped, 325 xfailed, 17 xpassed in 1093.55s (0:18:13) =
Tests should succeed to generate coverage reports using LCOV
```

**Comment by jorenham** [comment]
> Thanks @jorenham. This one's my fault. If I remember right, I had told @izaid I would add them in a follow-up to his Legendre function PR.
> 
> A couple things:
> 
> 1. We also need a deprecation warning for [`sph_harm`](https://scipy.github.io/devdocs/reference/generated/scipy.special.sph_harm.html#scipy.special.sph_harm), and suppression in the relevant tests.
> 2. Some tests in `test_mpmath.py` need to have the warning suppression added.
> 
> ```
>  =========================== short test summary info ============================
> FAILED scipy/special/tests/test_mpmath.py::TestSystematic::test_legenp - DeprecationWarning: `scipy.special.lpmn` is deprecated as of SciPy 1.15.0 and will be removed in SciPy 1.17.0. Please use `scipy.special.assoc_legendre_p_all` instead.
> FAILED scipy/special/tests/test_mpmath.py::TestSystematic::test_legenp_complex_2 - DeprecationWarning: `scipy.special.clpmn` is deprecated as of SciPy 1.15.0 and will be removed in SciPy 1.17.0. Please use `scipy.special.assoc_legendre_p_all` instead.
> FAILED scipy/special/tests/test_mpmath.py::TestSystematic::test_legenp_complex_3 - DeprecationWarning: `scipy.special.clpmn` is deprecated as of SciPy 1.15.0 and will be removed in SciPy 1.17.0. Please use `scipy.special.assoc_legendre_p_all` instead.
> = 3 failed, 74062 passed, 4634 skipped, 325 xfailed, 17 xpassed in 1093.55s (0:18:13) =
> Tests should succeed to generate coverage reports using LCOV
> ```

**Comment by jorenham** [comment]
woops, wrong button

**Comment by jorenham** [comment]
> * We also need a deprecation warning for [`sph_harm`](https://scipy.github.io/devdocs/reference/generated/scipy.special.sph_harm.html#scipy.special.sph_harm), and suppression in the relevant tests.

I'm not sure what the best way is to go about that, given that it's a ufunc 🤔

**Comment by jorenham** [comment]
FYI, the failing mypy tests are caused by a numpy 2.2.1 regression, for which aI opened https://github.com/numpy/numpy/pull/28053

**Comment by h-vetinari** [comment]
> FYI, the failing mypy tests are caused by a numpy 2.2.1 regression, for which aI opened [numpy/numpy#28053](https://github.com/numpy/numpy/pull/28053)

Can we skip the relevant tests here to get this merged? It would be good not to have to wait for numpy 2.2.2

**Comment by jorenham** [comment]
> > FYI, the failing mypy tests are caused by a numpy 2.2.1 regression, for which aI opened [numpy/numpy#28053](https://github.com/numpy/numpy/pull/28053)
> 
> Can we skip the relevant tests here to get this merged? It would be good not to have to wait for numpy 2.2.2

Is there anything I can do on my side to help with this?

FWIW, even though the mypy output is ridiculously huge, there are only 2 mypy errors if I'm not mistaken, so we could also temporarily `# type: ignore` those 🤷🏻

**Comment by h-vetinari** [comment]
> Is there anything I can do on my side to help with this?

I see you have maintainer (or at least triaging) rights for numpy - perhaps you can ask nicely if the numpy folks want to cut a 2.2.2 release. However, given that 2.2.1 came out ~24h ago, I kinda doubt this. So probably the `# type: ignore` approach here is more reasonable (and we can add an issue to revert that part as soon as 2.2.2 comes out).

**Comment by jorenham** [comment]
> I see you have maintainer (or at least triaging) rights for numpy - perhaps you can ask nicely if the numpy folks want to cut a 2.2.2 release.

Judging by https://github.com/numpy/numpy/pull/28055, I believe that @charris is already working on this.

**Comment by jorenham** [comment]
> > * We also need a deprecation warning for [`sph_harm`](https://scipy.github.io/devdocs/reference/generated/scipy.special.sph_harm.html#scipy.special.sph_harm), and suppression in the relevant tests.
> 
> I'm not sure what the best way is to go about that, given that it's a ufunc 🤔

I've thought about it for a bit now, and two ways come to mind:

- in `special/__init__.py` monkeypatch `sph_harm.__call__ = _deprecated(...)(sph_harm.__call__)`
- in `special/_ufuncs.pyi` replace `sph_harm: np.ufunc` with `sph_harm: _SphHarm`, define `class _SphHarm(np.ufunc): ...` and decorate it with `@typing.type_check_only` and `@typing_extensions.deprecated(...)`

**Comment by jorenham** [comment]
> So probably the `# type: ignore` approach here is more reasonable (and we can add an issue to revert that part as soon as 2.2.2 comes out).

https://github.com/scipy/scipy/pull/22162

**Comment by steppi** [comment]
> I'm not sure what the best way is to go about that, given that it's a ufunc 🤔

Ok, right, yeah. When `sph_harm` first got deprecated, I was thinking of putting the warning message in the scalar kernel like SciPy does here for a different function:

https://github.com/scipy/scipy/blob/6a5e3486be71524750f6fede26184cbd02ecea8f/scipy/special/_legacy.pxd#L45-L64

But thinking for a moment, it's actually not a very good idea for deprecating the function itself, because users could get an absolutely excessive amount of warnings.

It should go in the ufunc loop function before the looping starts. Don't worry about it; I'll take care of it later. There's going to be a very slick and clean way to do it with modern C++, and I'll have to confirm with @izaid about how he thinks it should be done.

https://github.com/scipy/scipy/blob/6a5e3486be71524750f6fede26184cbd02ecea8f/scipy/special/xsf/numpy.h#L808-L826

**Comment by h-vetinari** [comment]
> Judging by [numpy/numpy#28055](https://github.com/numpy/numpy/pull/28055), I believe that @charris is already working on this.

The only thing I see there is that it's milestoned for 2.2.2, which doesn't mean anything concrete about a release timeline. To unblock us here (and the 1.15 release), I think we should go with the `# type: ignore`. Thanks for https://github.com/scipy/scipy/pull/22162

**Comment by jorenham** [comment]
> But thinking for a moment, it's actually not a very good idea for deprecating the function itself, because users could get an absolutely excessive amount of warnings.

Ah yes I see the problem; we don't want to spam the user more than needed.

> It should go in the ufunc loop function before the looping starts. Don't worry about it; I'll take care of it later. There's going to be a very slick and clean way to do it with modern C++, and I'll have to confirm with @izaid about how he thinks it should be done.
> 
> https://github.com/scipy/scipy/blob/6a5e3486be71524750f6fede26184cbd02ecea8f/scipy/special/xsf/numpy.h#L808-L826

That sounds rather involved, hopefully it can make it in time before the 1.15 final release then.

---

FWIW, I'll `@deprecate` it in `scipy-stubs` (https://github.com/jorenham/scipy-stubs/issues/369) by using a similar trick as I described in that 2nd option.

**Comment by steppi** [comment]
> That sounds rather involved, hopefully it can make it in time before the 1.15 final release then.

I just looked into it a little more. It's not so bad at all because [PyErr_WarnEx](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_WarnEx) takes a stack level arg that controls where the warning is emitted, so it can just get called once instead of in each loop. That's actually what happens with `btdr`, I can submit a PR for it today.

**Comment by h-vetinari** [comment]
Could you please squash all the `TST:` commits into one? (in case you want to keep separate commits between the deprecation and the tests changes; otherwise we can squash-merge the entire PR).

File: scipy/special/tests/test_legendre.py
Comment (h-vetinari) [inline scipy/special/tests/test_legendre.py:697]: I don't know if or to what degree these tests were duplicated when introducing the new implementation, but if for some reason these tests are unique w.r.t. what aspect of `clpmn` resp. `assoc_legendre_p_all` they test, then we should simply update those tests to use the new function names, rather than sprinkling ignored warnings in lots of places.

**Review by h-vetinari — COMMENTED** [review]
Basically LGTM for merging, but would appreciate input from @steppi on the point below.

File: scipy/special/tests/test_legendre.py
Comment (jorenham) [reply scipy/special/tests/test_legendre.py:697]: Yea I think that makes sense, but it might be a bit out-of-scope to do this within this PR.

Comment (steppi) [reply scipy/special/tests/test_legendre.py:697]: Good thinking. What's going on though is that e.g. `assoc_legendre_p_all` isn't just a drop in replacement `clpmn`. The new functions provide a new, more featureful, and more consistent interface for Legendre functions.

`clpmn` calls `assoc_legendre_p_all` under the hood, but the process to convert to the right output type isn't exactly trivial:

https://github.com/scipy/scipy/blob/ba8553c12cfa4d3ee3df1f862230f360638dd64c/scipy/special/_basic.py#L1856-L1867

so we left the old tests in to make sure the conversion is correct. I think it's fine to keep them around until 1.17.0, but it would be worth having comments explaining why they are there.

Comment (h-vetinari) [reply scipy/special/tests/test_legendre.py:697]: OK, sounds good! The comments don't have to come as part of this PR.

**Review by h-vetinari — APPROVED** [review]

**Review by steppi — APPROVED** [review]
Looks good to me too. Many thanks @jorenham

**Pull request merged**
